### DISCUSSION TITLE
How to determine WebRTC dataChannel.send is delivered?

### DISCUSSION THREAD

**Discussion by jimmywarting** [pion/webrtc#2413] [Q&A] [answered]
I was searching for a way to determinate if a message have been delivered.
So i was investigating ACK message, and sending reply and doing some kind of RPC callback jank

but then i found this:
https://stackoverflow.com/questions/58256552/how-to-determine-webrtc-datachannel-send-is-delivered

> You may not need to, if you use an [ordered](https://w3c.github.io/webrtc-pc/#dom-datachannel-ordered), [reliable](https://w3c.github.io/webrtc-pc/#dfn-data-transport) datachannel (which is the default).
> 
> **An ordered reliable datachannel**
> With one of these, you determine a message was sent, by waiting for the [bufferedAmount](http://w3c.github.io/webrtc-pc/#dom-datachannel-bufferedamount) to go down:

I just want to be sure about one thing: if i use ordered and reliable, Will `bufferedamount` **only** decrease once the receiver have received the message?

meaning: it's not the underlying OS that actually sends the message over the network that decreases it... it's the receiver who says that?

**Comment by xenovation** [comment]
Just wanted to share some details regarding bufferedAmount if someone else came here from search engine, during my testing in chrome, for both ordered/unordered, reliable channel, after initiating a send, the bufferedAmount will increase and then immediately go down even if the data has not been sent yet, unless if the internal buffer of sctp is full (which is around 3MB), then bufferedAmount will start piling up.

**Reply by bhupiister** [reply]
Is this true? Did you dig deeper? 

This is the solution that most people are providing over the Internet. 

I also need an acknowledgment mechanism to determine that message was received.

**Reply by Sean-Der** [reply, marked as answer]
@bhupiister I would implement ACKs on top of SCTP if you care about an individual message. From MDN for bufferedAmount

```
The queue may build up as a result of calls to the [send()](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/send) method. This only includes data buffered by the user agent itself; it doesn't include any framing overhead or buffering done by the operating system or network hardware.
```
